File: crypto-manager.js
/**
 * Client-side encryption manager for What Do You Say? PWA
 * Uses Web Crypto API for secure encryption
 */

class PWACryptoManager {
  constructor() {
    this.authenticated = false;
    this.currentKey = null;
    this.masterKey = null;
    this.keyVersion = 1;
    this.iterations = 100000; // PBKDF2 iterations
  }

  /**
   * Initialize or authenticate with master password
   */
  async authenticate(masterPassword, setupNew = false) {
    try {
      if (setupNew) {
        return await this.setupFirstTime(masterPassword);
      } else {
        return await this.login(masterPassword);
      }
    } catch (error) {
      console.error('Authentication failed:', error);
      return false;
    }
  }

  /**
   * First-time setup with master password
   */
  async setupFirstTime(masterPassword) {
    try {
      console.log('🔐 Setting up client-side encryption...');

      // Generate salt for key derivation
      const salt = crypto.getRandomValues(new Uint8Array(32));
      
      // Derive key from password using PBKDF2
      const keyMaterial = await this.deriveKeyFromPassword(masterPassword, salt);
      
      // Generate additional encryption key
      const encryptionKey = await crypto.subtle.generateKey(
        { name: 'AES-GCM', length: 256 },
        true,
        ['encrypt', 'decrypt']
      );

      // Export the encryption key to encrypt it with the master key
      const exportedKey = await crypto.subtle.exportKey('raw', encryptionKey);
      
      // Encrypt the encryption key with the master key
      const iv = crypto.getRandomValues(new Uint8Array(12));
      const encryptedKey = await crypto.subtle.encrypt(
        { name: 'AES-GCM', iv: iv },
        keyMaterial,
        exportedKey
      );

      // Store encrypted key data
      const keyData = {
        salt: Array.from(salt),
        iv: Array.from(iv),
        encryptedKey: Array.from(new Uint8Array(encryptedKey)),
        keyVersion: this.keyVersion,
        created: new Date().toISOString(),
        keyRotationDue: new Date(Date.now() + 30 * 24 * 60 * 60 * 1000).toISOString() // 30 days
      };

      // Store in localStorage (encrypted)
      localStorage.setItem('whaddyasay_keydata', JSON.stringify(keyData));

      // Store auth info
      const authData = {
        lastAuth: new Date().toISOString(),
        authCount: 1,
        setupComplete: true
      };
      localStorage.setItem('whaddyasay_auth', JSON.stringify(authData));

      // Set current keys
      this.masterKey = keyMaterial;
      this.currentKey = encryptionKey;
      this.authenticated = true;

      console.log('✅ Client-side encryption setup complete!');
      return true;

    } catch (error) {
      console.error('❌ Setup failed:', error);
      return false;
    }
  }

  /**
   * Login with existing master password
   */
  async login(masterPassword) {
    try {
      const keyDataStr = localStorage.getItem('whaddyasay_keydata');
      if (!keyDataStr) {
        console.log('🔐 First time setup required');
        return await this.setupFirstTime(masterPassword);
      }

      const keyData = JSON.parse(keyDataStr);
      
      // Derive master key from password
      const salt = new Uint8Array(keyData.salt);
      const masterKey = await this.deriveKeyFromPassword(masterPassword, salt);

      // Decrypt the encryption key
      const iv = new Uint8Array(keyData.iv);
      const encryptedKey = new Uint8Array(keyData.encryptedKey);
      
      try {
        const decryptedKeyData = await crypto.subtle.decrypt(
          { name: 'AES-GCM', iv: iv },
          masterKey,
          encryptedKey
        );

        // Import the decrypted key
        const encryptionKey = await crypto.subtle.importKey(
          'raw',
          decryptedKeyData,
          { name: 'AES-GCM', length: 256 },
          true,
          ['encrypt', 'decrypt']
        );

        // Set current keys
        this.masterKey = masterKey;
        this.currentKey = encryptionKey;
        this.authenticated = true;

        // Update auth tracking
        this.updateAuthTracking();

        // Check if key rotation is due
        this.checkKeyRotation(keyData);

        console.log('✅ Authentication successful');
        return true;

      } catch (decryptError) {
        console.log('❌ Invalid password');
        return false;
      }

    } catch (error) {
      console.error('❌ Login failed:', error);
      return false;
    }
  }

  /**
   * Derive key from password using PBKDF2
   */
  async deriveKeyFromPassword(password, salt) {
    const encoder = new TextEncoder();
    const keyMaterial = await crypto.subtle.importKey(
      'raw',
      encoder.encode(password),
      { name: 'PBKDF2' },
      false,
      ['deriveBits', 'deriveKey']
    );

    return await crypto.subtle.deriveKey(
      {
        name: 'PBKDF2',
        salt: salt,
        iterations: this.iterations,
        hash: 'SHA-256'
      },
      keyMaterial,
      { name: 'AES-GCM', length: 256 },
      false,
      ['encrypt', 'decrypt']
    );
  }

  /**
   * Encrypt data
   */
  async encryptData(data) {
    if (!this.authenticated || !this.currentKey) {
      throw new Error('Not authenticated - call authenticate() first');
    }

    try {
      const encoder = new TextEncoder();
      const jsonData = JSON.stringify(data);
      const dataBuffer = encoder.encode(jsonData);

      const iv = crypto.getRandomValues(new Uint8Array(12));
      const encryptedData = await crypto.subtle.encrypt(
        { name: 'AES-GCM', iv: iv },
        this.currentKey,
        dataBuffer
      );

      // Combine IV and encrypted data
      const result = new Uint8Array(iv.length + encryptedData.byteLength);
      result.set(iv);
      result.set(new Uint8Array(encryptedData), iv.length);

      // Return as base64 string
      return btoa(String.fromCharCode(...result));

    } catch (error) {
      throw new Error(`Encryption failed: ${error.message}`);
    }
  }

  /**
   * Decrypt data
   */
  async decryptData(encryptedDataStr) {
    if (!this.authenticated || !this.currentKey) {
      throw new Error('Not authenticated - call authenticate() first');
    }

    try {
      // Convert from base64
      const encryptedData = new Uint8Array(
        atob(encryptedDataStr).split('').map(c => c.charCodeAt(0))
      );

      // Extract IV and encrypted data
      const iv = encryptedData.slice(0, 12);
      const data = encryptedData.slice(12);

      // Decrypt
      const decryptedData = await crypto.subtle.decrypt(
        { name: 'AES-GCM', iv: iv },
        this.currentKey,
        data
      );

      // Convert back to JSON
      const decoder = new TextDecoder();
      const jsonString = decoder.decode(decryptedData);
      return JSON.parse(jsonString);

    } catch (error) {
      throw new Error(`Decryption failed: ${error.message}`);
    }
  }

  /**
   * Rotate encryption keys
   */
  async rotateKeys(currentPassword, newPassword = null) {
    try {
      console.log('🔄 Rotating encryption keys...');

      if (!await this.login(currentPassword)) {
        return false;
      }

      const passwordToUse = newPassword || currentPassword;

      // Create backup of current keys
      const currentKeyData = localStorage.getItem('whaddyasay_keydata');
      const backupKey = `whaddyasay_backup_${Date.now()}`;
      localStorage.setItem(backupKey, currentKeyData);

      // Set up new keys
      const success = await this.setupFirstTime(passwordToUse);

      if (success) {
        console.log('✅ Key rotation complete');
        console.log('🔒 Old keys backed up securely');
        return true;
      } else {
        console.log('❌ Key rotation failed');
        return false;
      }

    } catch (error) {
      console.error('❌ Key rotation error:', error);
      return false;
    }
  }

  /**
   * Update authentication tracking
   */
  updateAuthTracking() {
    try {
      const authData = JSON.parse(localStorage.getItem('whaddyasay_auth') || '{}');
      authData.lastAuth = new Date().toISOString();
      authData.authCount = (authData.authCount || 0) + 1;
      localStorage.setItem('whaddyasay_auth', JSON.stringify(authData));
    } catch (error) {
      console.warn('Could not update auth tracking:', error);
    }
  }

  /**
   * Check if key rotation is due
   */
  checkKeyRotation(keyData) {
    try {
      const rotationDue = new Date(keyData.keyRotationDue);
      if (new Date() > rotationDue) {
        console.warn('⚠️  Key rotation recommended!');
        console.warn('   Your encryption keys are over 30 days old');
        console.warn('   Consider rotating keys for better security');
      }
    } catch (error) {
      // Non-critical
    }
  }

  /**
   * Get current security status
   */
  getSecurityStatus() {
    const status = {
      authenticated: this.authenticated,
      setupComplete: false,
      lastAuth: null,
      authCount: 0,
      keyRotationDue: null,
      daysUntilRotation: null
    };

    try {
      // Get auth data
      const authData = JSON.parse(localStorage.getItem('whaddyasay_auth') || '{}');
      Object.assign(status, authData);

      // Get key rotation info
      const keyData = JSON.parse(localStorage.getItem('whaddyasay_keydata') || '{}');
      if (keyData.keyRotationDue) {
        status.keyRotationDue = keyData.keyRotationDue;
        const rotationDate = new Date(keyData.keyRotationDue);
        const daysUntil = Math.ceil((rotationDate - new Date()) / (1000 * 60 * 60 * 24));
        status.daysUntilRotation = daysUntil;
      }

    } catch (error) {
      console.error('Error getting security status:', error);
    }

    return status;
  }

  /**
   * Clear all stored data (logout)
   */
  logout() {
    this.authenticated = false;
    this.currentKey = null;
    this.masterKey = null;
    console.log('🔐 Logged out - encryption keys cleared from memory');
  }

  /**
   * Generate secure random password
   */
  generateSecurePassword(length = 16) {
    const charset = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789!@#$%^&*';
    const array = new Uint8Array(length);
    crypto.getRandomValues(array);
    return Array.from(array, byte => charset[byte % charset.length]).join('');
  }

  /**
   * Test encryption/decryption
   */
  async testEncryption() {
    if (!this.authenticated) {
      throw new Error('Not authenticated');
    }

    const testData = {
      message: 'Hello, World!',
      timestamp: new Date().toISOString(),
      number: 42
    };

    try {
      const encrypted = await this.encryptData(testData);
      const decrypted = await this.decryptData(encrypted);
      
      const success = JSON.stringify(testData) === JSON.stringify(decrypted);
      console.log('🧪 Encryption test:', success ? 'PASSED' : 'FAILED');
      return success;
    } catch (error) {
      console.error('🧪 Encryption test FAILED:', error);
      return false;
    }
  }
}

// Export for use in other modules
window.PWACryptoManager = PWACryptoManager;

File: crypto_manager.py
#!/usr/bin/env python3
"""
Personal Data Encryption Manager
Provides end-to-end encryption for personal conversation data
User owns and controls all encryption keys
"""

import os
import json
import base64
import hashlib
import secrets
from datetime import datetime, timedelta
from pathlib import Path
from typing import Optional, Dict, Any, Tuple
from cryptography.fernet import Fernet
from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC
from cryptography.hazmat.primitives.asymmetric import rsa, padding
import getpass

class PersonalCryptoManager:
    """
    Manages encryption for personal conversation data
    - User-controlled master password
    - Monthly key rotation (recommended)
    - Local key storage only
    - No cloud dependencies
    """
    
    def __init__(self, data_dir: str = "./data"):
        self.data_dir = Path(data_dir)
        self.data_dir.mkdir(exist_ok=True)
        self.keys_dir = self.data_dir / "keys"
        self.keys_dir.mkdir(exist_ok=True, mode=0o700)  # Restricted permissions
        
        self.master_key_file = self.keys_dir / "master.key"
        self.auth_file = self.keys_dir / "auth.json"
        self.current_key = None
        self.authenticated = False
        
    def setup_first_time(self, master_password: str) -> bool:
        """
        First-time setup: create master key and authentication
        """
        try:
            print("🔐 Setting up your personal encryption...")
            
            # Generate salt for password derivation
            salt = os.urandom(32)
            
            # Derive key from password
            kdf = PBKDF2HMAC(
                algorithm=hashes.SHA256(),
                length=32,
                salt=salt,
                iterations=100000,
            )
            key = base64.urlsafe_b64encode(kdf.derive(master_password.encode()))
            
            # Create Fernet cipher
            fernet = Fernet(key)
            
            # Generate RSA key pair for additional security
            private_key = rsa.generate_private_key(
                public_exponent=65537,
                key_size=2048,
            )
            
            # Serialize keys
            private_pem = private_key.private_bytes(
                encoding=serialization.Encoding.PEM,
                format=serialization.PrivateFormat.PKCS8,
                encryption_algorithm=serialization.NoEncryption()
            )
            
            public_key = private_key.public_key()
            public_pem = public_key.public_bytes(
                encoding=serialization.Encoding.PEM,
                format=serialization.PublicFormat.SubjectPublicKeyInfo
            )
            
            # Encrypt the private key with Fernet
            encrypted_private_key = fernet.encrypt(private_pem)
            
            # Store master key info
            master_data = {
                "salt": base64.b64encode(salt).decode(),
                "encrypted_private_key": base64.b64encode(encrypted_private_key).decode(),
                "public_key": base64.b64encode(public_pem).decode(),
                "created_at": datetime.now().isoformat(),
                "key_rotation_due": (datetime.now() + timedelta(days=30)).isoformat()
            }
            
            with open(self.master_key_file, 'w') as f:
                json.dump(master_data, f, indent=2)
            
            # Set restrictive permissions
            os.chmod(self.master_key_file, 0o600)
            
            # Create auth tracking
            auth_data = {
                "last_auth": datetime.now().isoformat(),
                "auth_count": 1,
                "setup_complete": True
            }
            
            with open(self.auth_file, 'w') as f:
                json.dump(auth_data, f, indent=2)
            
            os.chmod(self.auth_file, 0o600)
            
            self.current_key = fernet
            self.authenticated = True
            
            print("✅ Personal encryption setup complete!")
            print("🔑 Your data is now encrypted and only you have the key")
            print("📅 Key rotation recommended in 30 days")
            
            return True
            
        except Exception as e:
            print(f"❌ Setup failed: {e}")
            return False
    
    def authenticate(self, master_password: str) -> bool:
        """
        Authenticate user and load encryption keys
        """
        try:
            if not self.master_key_file.exists():
                print("🔐 First time setup required")
                return self.setup_first_time(master_password)
            
            # Load master key data
            with open(self.master_key_file, 'r') as f:
                master_data = json.load(f)
            
            # Derive key from password
            salt = base64.b64decode(master_data["salt"])
            kdf = PBKDF2HMAC(
                algorithm=hashes.SHA256(),
                length=32,
                salt=salt,
                iterations=100000,
            )
            
            try:
                key = base64.urlsafe_b64encode(kdf.derive(master_password.encode()))
                fernet = Fernet(key)
                
                # Test decryption to verify password
                encrypted_private_key = base64.b64decode(master_data["encrypted_private_key"])
                private_pem = fernet.decrypt(encrypted_private_key)
                
                # If we get here, password is correct
                self.current_key = fernet
                self.authenticated = True
                
                # Update auth tracking
                self._update_auth_tracking()
                
                # Check if key rotation is due
                self._check_key_rotation(master_data)
                
                print("✅ Authentication successful")
                return True
                
            except Exception:
                print("❌ Invalid password")
                return False
                
        except Exception as e:
            print(f"❌ Authentication failed: {e}")
            return False
    
    def encrypt_data(self, data: Dict[Any, Any]) -> str:
        """
        Encrypt personal data
        """
        if not self.authenticated:
            raise Exception("Not authenticated - call authenticate() first")
        
        # Convert to JSON and encrypt
        json_data = json.dumps(data, default=str)
        encrypted_data = self.current_key.encrypt(json_data.encode())
        
        # Return base64 encoded for storage
        return base64.b64encode(encrypted_data).decode()
    
    def decrypt_data(self, encrypted_data: str) -> Dict[Any, Any]:
        """
        Decrypt personal data
        """
        if not self.authenticated:
            raise Exception("Not authenticated - call authenticate() first")
        
        try:
            # Decode and decrypt
            encrypted_bytes = base64.b64decode(encrypted_data)
            decrypted_bytes = self.current_key.decrypt(encrypted_bytes)
            
            # Parse JSON
            return json.loads(decrypted_bytes.decode())
            
        except Exception as e:
            raise Exception(f"Decryption failed: {e}")
    
    def rotate_keys(self, master_password: str, new_password: Optional[str] = None) -> bool:
        """
        Rotate encryption keys (recommended monthly)
        """
        try:
            print("🔄 Rotating encryption keys...")
            
            if not self.authenticate(master_password):
                return False
            
            # Use new password if provided, otherwise keep current
            password_to_use = new_password if new_password else master_password
            
            # Create backup of current keys
            backup_file = self.keys_dir / f"backup_{datetime.now().strftime('%Y%m%d_%H%M%S')}.key"
            if self.master_key_file.exists():
                import shutil
                shutil.copy2(self.master_key_file, backup_file)
                os.chmod(backup_file, 0o600)
            
            # Generate new keys
            success = self.setup_first_time(password_to_use)
            
            if success:
                print("✅ Key rotation complete")
                print("🔒 Old keys backed up securely")
                return True
            else:
                print("❌ Key rotation failed")
                return False
                
        except Exception as e:
            print(f"❌ Key rotation error: {e}")
            return False
    
    def _update_auth_tracking(self):
        """Update authentication tracking"""
        try:
            auth_data = {}
            if self.auth_file.exists():
                with open(self.auth_file, 'r') as f:
                    auth_data = json.load(f)
            
            auth_data.update({
                "last_auth": datetime.now().isoformat(),
                "auth_count": auth_data.get("auth_count", 0) + 1
            })
            
            with open(self.auth_file, 'w') as f:
                json.dump(auth_data, f, indent=2)
                
        except Exception as e:
            print(f"Warning: Could not update auth tracking: {e}")
    
    def _check_key_rotation(self, master_data: Dict):
        """Check if key rotation is recommended"""
        try:
            rotation_due = datetime.fromisoformat(master_data["key_rotation_due"])
            if datetime.now() > rotation_due:
                print("⚠️  Key rotation recommended!")
                print("   Your encryption keys are over 30 days old")
                print("   Consider running: crypto_manager.rotate_keys()")
                
        except Exception:
            pass  # Non-critical
    
    def get_security_status(self) -> Dict[str, Any]:
        """Get current security status"""
        status = {
            "authenticated": self.authenticated,
            "setup_complete": False,
            "last_auth": None,
            "auth_count": 0,
            "key_rotation_due": None,
            "days_until_rotation": None
        }
        
        try:
            if self.auth_file.exists():
                with open(self.auth_file, 'r') as f:
                    auth_data = json.load(f)
                status.update(auth_data)
            
            if self.master_key_file.exists():
                with open(self.master_key_file, 'r') as f:
                    master_data = json.load(f)
                
                rotation_due = datetime.fromisoformat(master_data["key_rotation_due"])
                status["key_rotation_due"] = master_data["key_rotation_due"]
                status["days_until_rotation"] = (rotation_due - datetime.now()).days
                
        except Exception:
            pass
        
        return status